bpo-29757: don't swallows errors in the `socket.create_connection()` utility loop

ankostis opened the pull request:
## Context
The utility method `socket.create_connection()` currently works like that:
1. resolve the destination-address into one or more IP(v4 & v6) addresses;
2. loop on each IP address and stop to the 1st one to work;
3. if none works, re-raise the last error.

## The problem
So currently the loop in `socket.create_connection()` ignores all intermediate errors and reports only the last connection failure, which  might be irrelevant.  For instance, when both IPv4 & IPv6 networks are supported, usually the last address is a IPv6 address and it frequently fails with an irrelevant error - the actual cause have already been ignored.

## Possible solutions
To facilitate network debugging, there are at least 3 options:

a. log each failure [as it happens](/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Lib/socket.py#L717), but that would get the final failure twice: once as a (warning?) message, and once as an exception .
b. collect all failures and log them only when connection fails, but that might miss important infos to the user;
c. collect and return all failures in list attached to the raised exception, which also might miss important infos.

### open questions
Q1. A question for all cases is what logging "means" to use: the `warnings` or `logging` module?
Q2. And if `logging` is chosen, log them in `'DEBUG'` or `'WARNING'` level?
Q3. Finally, cases (b) & (c) need an answer whether to log collected errors in the case that connection passes.

The 1st part of case (c) is implemented in this PR.

ankostis committed:
feat(socket): don't swallows errors in `socket.create_connection()` loop

## Context
The utility method `socket.create_connection()` currently works like that:
1. resolve the destination-address into one or more IP(v4 & v6) addresses;
2. loop on each IP address and stop to the 1st one to work;
3. if none works, re-raise the last error.

## The problem
So currently the loop in `socket.create_connection()` ignores all intermediate errors and reports only the last connection failure, 
which  might be irrelevant.  For instance, when both IPv4 & IPv6 networks are supported, usually the last address is a IPv6 address and it frequently fails with an irrelevant error - the actual cause have already been ignored.

## Possible solutions & open questions
To facilitate network debugging, there are at least 3 options:
a. log each failure [as they happen](/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Lib/socket.py#L717), but that would get the final failure twice: once as a (warning?) message, and once as an exception .
b. collect all failures and log them only when connection fails to collect the errors, 
   but that might miss important infos to the user;
c. collect and return all failures in list attached to the raised exception.

A question for cases (a) & (b) is what logging "means" to use: the `warnings` or `logging` module?
And if `logging` is chosen, log them in `'DEBUG'` or `'WARNING'` level?

Case (c), implemented in this PR sidesteps the above questions.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA. This is necessary for legal reasons before we can look at your contribution. Please follow these steps to help rectify the issue:

1. If you don't have an account on b.p.o, please [create one](http://bugs.python.org/user?@template=register)
2. Make sure your GitHub username is listed in ["Your Details" at b.p.o](https://cloud.githubusercontent.com/assets/2680980/23276970/d14a380c-f9d1-11e6-883d-e13b6b211239.png)
3. If you have not already done so, please sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/). The "bugs.python.org username " requested by the form is the "Login name" field under "Your Details".
4. If you just signed the CLA, please **wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
5. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

brettcannon commented:
Could you open an issue to discuss this, @ankostis ? We prefer to discuss design and such on bugs.python.org so that GitHub is only used for code review.

ankostis commented:
[Issue 29757](http://bugs.python.org/issue29757).

ankostis commented:
By the way my CLA has been signed.

brettcannon commented:
Thanks. And just in case you didn't notice, Travis failed (not sure if it's because of the PR changes).

ankostis commented:
Thanks, I will look at it tomorrow.

brettcannon commented:
@ankostis you can also wait for any discussion to be had about the issue before you put in more time in case this isn't accepted for some reason.

ankostis commented:
@brettcannon  grateful for your care to contributors like me. 

Anyhow, I did some changes on the OP regarding what I tried to implement and what I consider as possible solutions & open questions.

1st1 commented on Lib/socket.py:
```diff
@@ -714,12 +714,12 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
             return sock
 
         except error as _:
```
Please don't use `_`, it would look much better with a normal variable name such as `e`.

vstinner replied:
Please don't use one letter name like "e" :-) Use "error", "exc" or whatever else.

ankostis replied:
Done in ~~ec887c0c3~~  (rebased on latest master).
Had to rewrite again (b39d4b1c) to rename `e` --> `ex` (i like it also this way, just trying to preserve styles).

1st1 commented on Lib/socket.py:
```diff
@@ -700,7 +700,7 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
 
     host, port = address
-    err = None
+    errs = []
```
`errs` -> `errors`

ankostis replied:
Done in ~~ec887c0c3~~ b39d4b1c (rebased on latest master).

1st1 commented:
I'll post some feedback to bpo.

ankostis committed:
refact(socket): corrections according to #562

- Rename `_` and `errs` variable names.
- Raise original error if a single one.
- Improve a bit error message, but still no Errno provided for when
multiple errors are included.

ankostis commented:
> And just in case you didn't notice, Travis failed (not sure if it's because of the PR changes).

Now all TCs pass, because I re-raise the exact original error in case only a single one was collected.

But still, this PR is needs more work because of 2 issues: 

### Abuse of `OSError()` constructor
When more than one error is collected, I pass a string to the 1st arg to OSError() constructor  (instead of an `errno`).
But I copied this from [existing code](https://github.com/python/cpython/blob/master/Lib/socket.py#L724).

Which `errno` should use there?


### Notify user about collected errors on Success
The list of intermediate errors is lost in the case of success.
As suggested on the OP, this ~~may happen~~ can be corrected (in my order of preference):

1. with a new optional argument for the user to provide the list to collect the errors (it changes the API backward-compatibly but allows inspection of the errors);
2. with logging logs (more configurable than warning, but no inspection - only strings);
3. with warnings;
4. by doing nothing (swallowing error-messages is dangerous in the long term).

ankostis commented:
@the-knights-who-say-ni  I have signed CLA; is there some problem with it?

brettcannon commented:
@ankostis I just checked the logs and it seems you didn't set your GitHub username on your bugs.python.org (i.e. the bot can't find any account on bugs.python.org that has the GitHub username "ankostis" set). Fix that and hopefully the bot will pick up that you signed the CLA.

ankostis commented:
OK, my gh-username has been set.

brettcannon commented:
@ankostis that worked!

ankostis commented:
Any ideas [which road](https://github.com/python/cpython/pull/562#issuecomment-287419089) to take?

And where should the discussion take place, here on on [python-tracker](http://bugs.python.org/issue29757)?

brettcannon commented:
@ankostis code review issues are discussed here, all other discussions (such as what the best approach is, etc.) should happen on the issue.

MaxwellDupre reviewed (commented):
I see that PEP 654 -- Exception Groups and except*
would be the best way to implement multiple exceptions here, but this will need to wait while being implemented. Or you could try the without except* 
https://www.python.org/dev/peps/pep-0654/#programming-without-except
